File: cpp/source/CodeGenerator.cpp
#include "pch.h"
#include "CodeGenerator.h"
#include "../packages/utilities/Utilities.h"

const int TOTAL_FILES = 25;
const std::string INPUT_PATH = "./source/inputs/";
const std::string SAMPLES_PATH = "./source/samples/";
const std::string PROBLEMS_PATH = "./source/problems/";
const std::string SOLUTIONS_PATH = "./source/solutions/";

namespace CodeGenerator
{

	void GenerateMarkDownFiles()
	{
		std::filesystem::create_directories(PROBLEMS_PATH);
		auto lines = std::vector<std::string>{ "" };
		for (auto i = 1; i <= TOTAL_FILES; ++i)
		{
			for (auto j = 0; j < 2; ++j)
			{
				char part = (j == 0 ? 'A' : 'B');
				auto builder = std::stringstream{};
				builder << PROBLEMS_PATH << "Problem" << std::setfill('0') << std::setw(2) << i << part << ".md";
				Utilities::string_vector_vector_to_file(builder.str(), lines);
			}
		}
	}

	void GenerateHeaders()
	{
		auto lines = std::vector<std::string>{ "#pragma once" };

		for (auto i = 1; i <= TOTAL_FILES; ++i)
		{
			for (auto j = 0; j < 2; ++j)
			{
				const auto suffix = j == 0 ? 'A' : 'B';

				lines.emplace_back("");

				auto builder = std::stringstream{};
				builder << "namespace Solutions" << std::setfill('0') << std::setw(2) << i << suffix;
				lines.emplace_back(builder.str());
				lines.emplace_back("{");
				lines.emplace_back("\tvoid PrintSolution(const std::filesystem::path& inputFile, bool shouldRender);");
				lines.emplace_back("}");
			}
		}

		Utilities::string_vector_vector_to_file("source/solutions/Solutions.h", lines);
	}

	void GenerateSolutions()
	{
		std::filesystem::create_directories(SOLUTIONS_PATH);

		// line # 5, 13
		std::string line5 = "namespace Solutions";
		std::string line13 = "} // namespace Solutions";

		std::ifstream input_file("./build/debug/SolutionTemplate.txt");
		std::vector<std::string> lines;
		std::string input;

		while (std::getline(input_file, input))
		{
			lines.push_back(input);
			std::cout << input << std::endl;
		}
		input_file.close();

		for (auto i = 1; i <= TOTAL_FILES; i++)
		{
			for (auto j = 0; j < 2; ++j)
			{
				char part = (j == 0 ? 'A' : 'B');
				std::string day = (i < 10 ? '0' + std::to_string(i) : std::to_string(i));
				std::string output_file =  SOLUTIONS_PATH + "Solution" + day + part + ".cpp";

				lines[4] = line5 + day + part;
				lines[12] = line13 + day + part;

				std::cout << lines[12] << std::endl;
				Utilities::string_vector_vector_to_file(output_file, lines);
			}
		}
	}

	void GenerateInputFiles()
	{
		std::filesystem::create_directories(INPUT_PATH);
		std::filesystem::create_directories(SAMPLES_PATH);
		auto lines = std::vector<std::string>{ "" };
		for (auto i = 1; i <= TOTAL_FILES; ++i)
		{
			{
				auto builder = std::stringstream{};
				builder << INPUT_PATH << "Problem" << std::setfill('0') << std::setw(2) << i << ".input";
				Utilities::string_vector_vector_to_file(builder.str(), lines);
			}

			{
				auto builder = std::stringstream{};
				builder << SAMPLES_PATH << "Sample" << std::setfill('0') << std::setw(2) << i << ".input";
				Utilities::string_vector_vector_to_file(builder.str(), lines);
			}
		}
	}
} // namespace CodeGenerator

File: go/source/cmd/CodeGenerator.go
package main

import (
	"fmt"
	"log"
	"os"
	"strconv"
)

const FILE_START = 1
const FILE_END = 25
const PUZZLES = "/source/problems/"
const SOLUTIONS = "/source/solutions/"
const INPUTS = "/source/inputs/"
const SAMPLES = "/source/samples/"
const MARKDOWN = ".md"
const INPUT = ".input"
const GO = ".go"

func CodeGenerator() {
	GenerateMarkdownFiles()
	GenerateSolutions()
	GenerateInputFiles()
	GenerateSampleInputFiles()
}

func GenerateMarkdownFiles() {
	for FILE_START := 1; FILE_START <= FILE_END; FILE_START++ {
		for index := 0; index <= 1; index++ {

			stringBuilder, err := os.Getwd()
			if err != nil {
				log.Fatal(err)
			}
			stringBuilder += PUZZLES
			if _, err := os.Stat(stringBuilder); os.IsNotExist(err) {
				err = os.Mkdir(stringBuilder, os.ModeDir)
				if err != nil {
					log.Fatal(err)
				}
			}
			stringBuilder += "Problem"
			if FILE_START < 10 {
				stringBuilder += "0"
			}
			stringBuilder += strconv.Itoa(FILE_START)
			if index == 0 {
				stringBuilder += "A"
			}
			if index == 1 {
				stringBuilder += "B"
			}
			stringBuilder += MARKDOWN
			_, err = os.Create(stringBuilder)
			if err != nil {
				log.Fatal(err)
			}
		}
	}
}

func GenerateSolutions() {
	for FILE_START := 1; FILE_START <= FILE_END; FILE_START++ {
		for index := 0; index <= 1; index++ {
			directoryBuilder, err := os.Getwd()
			if err != nil {
				log.Fatal(err)
			}
			directoryBuilder += SOLUTIONS
			if _, err := os.Stat(directoryBuilder); os.IsNotExist(err) {
				if err = os.Mkdir(directoryBuilder, os.ModeDir); err != nil {
					log.Fatal(err)
				}
			}

			fileBuilder := "Solution"
			if FILE_START < 10 {
				fileBuilder += "0"
			}
			fileBuilder += strconv.Itoa(FILE_START)
			template := ""
			if index == 0 {
				fileBuilder += "A"
				template = fmt.Sprintf(SOLUTION_TEMPLATE, fileBuilder)
			}
			if index == 1 {
				fileBuilder += "B"
				template = fmt.Sprintf(SOLUTION_TEMPLATE, fileBuilder)
			}
			fileBuilder += GO
			file, err := os.Create(directoryBuilder + fileBuilder)
			if err != nil {
				log.Fatal(err)
			}
			if _, err := file.Write([]byte(template)); err != nil {
				log.Fatal(err)
			}
			if err := file.Close(); err != nil {
				log.Fatal(err)
			}
		}
	}
}

func GenerateInputFiles() {
	for FILE_START := 1; FILE_START <= FILE_END; FILE_START++ {
		stringBuilder, err := os.Getwd()
		if err != nil {
			log.Fatal(err)
		}
		stringBuilder += INPUTS
		if _, err := os.Stat(stringBuilder); os.IsNotExist(err) {
			err = os.Mkdir(stringBuilder, os.ModeDir)
			if err != nil {
				log.Fatal(err)
			}
		}
		stringBuilder += "Input"
		if FILE_START < 10 {
			stringBuilder += "0"
		}
		stringBuilder += strconv.Itoa(FILE_START)
		stringBuilder += INPUT
		_, err = os.Create(stringBuilder)
		if err != nil {
			log.Fatal(err)
		}
	}
}

func GenerateSampleInputFiles() {
	for FILE_START := 1; FILE_START <= FILE_END; FILE_START++ {
		stringBuilder, err := os.Getwd()
		if err != nil {
			log.Fatal(err)
		}
		stringBuilder += SAMPLES
		if _, err := os.Stat(stringBuilder); os.IsNotExist(err) {
			err = os.Mkdir(stringBuilder, os.ModeDir)
			if err != nil {
				log.Fatal(err)
			}
		}
		stringBuilder += "Sample"
		if FILE_START < 10 {
			stringBuilder += "0"
		}
		stringBuilder += strconv.Itoa(FILE_START)
		stringBuilder += INPUT
		_, err = os.Create(stringBuilder)
		if err != nil {
			log.Fatal(err)
		}
	}
}

// if _, err := os.Stat("/path/to/whatever"); os.IsNotExist(err) {
//     // path/to/whatever does not exist
// }

// if _, err := os.Stat("/path/to/whatever"); err == nil {
//     // path/to/whatever exists
// }
